sklearn/linear_model/_linear_loss.py

```diff
@@ -0,0 +1,362 @@
+            Sample weights.
+        alpha: float
+            L2 regularization strength
+        n_threads : int, default=1
+            Number of OpenMP threads to use.
+
+        Returns
+        -------
+        gradient : ndarray of shape (n_dof,) or (n_classes * n_dof)
+             The gradient of the loss as ravelled array.
+
+        hessp : callable
+            Function that takes in a vector input of shape of gradient and
+            and returns matrix-vector product with hessian.
+        """
+        (n_samples, n_features), n_classes = X.shape, self._loss.n_classes
+        w, intercept, raw_prediction = self._w_intercept_raw(coef, X)
+
+        if not self._loss.is_multiclass:
+            gradient, hessian = self._loss.gradient_hessian(
+                y_true=y,
+                raw_prediction=raw_prediction,
+                sample_weight=sample_weight,
+                n_threads=n_threads,
+            )
+            grad = np.empty_like(coef, dtype=X.dtype)
+            grad[:n_features] = X.T @ gradient + alpha * w
+            if self.fit_intercept:
+                grad[-1] = gradient.sum()
+
+            # Precompute as much as possible: hX, hh_intercept and hsum
+            hsum = hessian.sum()
+            if sparse.issparse(X):
+                hX = sparse.dia_matrix((hessian, 0), shape=(n_samples, n_samples)) @ X
+            else:
+                hX = hessian[:, np.newaxis] * X
+
+            if self.fit_intercept:
+                # Calculate the double derivative with respect to intercept.
+                # Note: In case hX is sparse, hX.sum is a matrix object.
+                hh_intercept = np.squeeze(np.array(hX.sum(axis=0)))
+
+            # With intercept included and alpha = 0, hessp returns
+            # res = (X, 1)' @ diag(h) @ (X, 1) @ s
+            #     = (X, 1)' @ (hX @ s[:n_features], sum(h) * s[-1])
+            # res[:n_features] = X' @ hX @ s[:n_features] + sum(h) * s[-1]
+            # res[:-1] = 1' @ hX @ s[:n_features] + sum(h) * s[-1]
+            def hessp(s):
+                ret = np.empty_like(s)
+                if sparse.issparse(X):
+                    ret[:n_features] = X.T @ (hX @ s[:n_features])
+                else:
+                    ret[:n_features] = np.linalg.multi_dot([X.T, hX, s[:n_features]])
+                ret[:n_features] += alpha * s[:n_features]
+
+                if self.fit_intercept:
+                    ret[:n_features] += s[-1] * hh_intercept
+                    ret[-1] = hh_intercept @ s[:n_features] + hsum * s[-1]
+                return ret
+
+        else:
+            # Here we may safely assume HalfMultinomialLoss aka categorical
+            # cross-entropy.
+            # HalfMultinomialLoss computes only the diagonal part of the hessian, i.e.
+            # diagonal in the classes. Here, we want the matrix-vector product of the
+            # full hessian. Therefore, we call gradient_proba.
```

Comment:
This is really special for multinomial case / `n_classes >= 3`. The hessian returned by HalfMultinomialLoss is only the diagonal part, diagonal in classes. Therefore, we call the method `gradient_proba` which exists only for HalfMultinomialLoss for exactly this reason: to define `hessp` such that it computes products of a vector with the **full hessian** for which gradient and predicted probabilties are enough.